core/primitives/src/validator_mandates/mod.rs

```diff
@@ -102,95 +98,105 @@ impl ValidatorMandates {
-    /// [`ShuffledShardIds`].
-    ///
-    /// It clones mandates since [`ValidatorMandates`] is supposed to be valid for an epoch, while a
-    /// new assignment is calculated at every height.
-    pub fn sample<R>(&self, rng: &mut R) -> ChunkValidatorStakeAssignment
-    where
-        R: Rng + ?Sized,
-    {
-        // Shuffling shard ids to avoid a bias towards lower ids, see [`ShuffledShardIds`]. We
-        // do two separate shuffes for full and partial mandates to reduce the likelihood of
-        // assigning fewer full _and_ partial mandates to the _same_ shard.
-        let shard_ids_for_mandates = ShuffledShardIds::new(rng, self.config.num_shards);
-        let shard_ids_for_partials = ShuffledShardIds::new(rng, self.config.num_shards);
-
-        let shuffled_mandates = self.shuffled_mandates(rng);
-        let shuffled_partials = self.shuffled_partials(rng);
-
-        // Distribute shuffled mandates and partials across shards. For each shard with `shard_id`
-        // in `[0, num_shards)`, we take the elements of the vector with index `i` such that `i %
-        // num_shards == shard_id`.
-        //
-        // Assume, for example, there are 10 mandates and 4 shards. Then for `shard_id = 1` we
-        // collect the mandates with indices 1, 5, and 9.
-        let stake_per_mandate = self.stake_per_mandate;
-        let mut stake_assignment_per_shard = vec![HashMap::new(); self.config.num_shards];
-        for shard_id in 0..self.config.num_shards {
-            // Achieve shard id shuffling by writing to the position of the alias of `shard_id`.
-            let mandates_assignment =
-                &mut stake_assignment_per_shard[shard_ids_for_mandates.get_alias(shard_id)];
-
-            // For the current `shard_id`, collect mandates with index `i` such that
-            // `i % num_shards == shard_id`.
-            for idx in (shard_id..shuffled_mandates.len()).step_by(self.config.num_shards) {
-                let validator_id = shuffled_mandates[idx];
-                *mandates_assignment.entry(validator_id).or_default() += stake_per_mandate;
+#[cfg(feature = "rand")]
+mod validator_mandates_sample {
+    use super::*;
+    use itertools::Itertools;
+    use rand::{seq::SliceRandom, Rng};
+
+    impl ValidatorMandates {
+        /// Returns a validator assignment obtained by shuffling mandates and assigning them to shards.
+        /// Shard ids are shuffled as well in this process to avoid a bias lower shard ids, see
+        /// [`ShuffledShardIds`].
+        ///
+        /// It clones mandates since [`ValidatorMandates`] is supposed to be valid for an epoch, while a
+        /// new assignment is calculated at every height.
+        pub fn sample<R>(&self, rng: &mut R) -> ChunkValidatorStakeAssignment
+        where
+            R: Rng + ?Sized,
+        {
+            // Shuffling shard ids to avoid a bias towards lower ids, see [`ShuffledShardIds`]. We
```

Comment:
Agree. Though not sure how to prevent that ... My only thought that code review + eventually we could make primitives quite thin and limit by amount of deps 👍🏻